Add ReaderFrom/WriterTo for Linux (splice)

cpuguy83 opened the pull request:
On Linux, we can use `splice` to optimize copies to happen in kernel
when copying to other files.
Since this is already a pipe, this should always work when copying
to/from another file.

cpuguy83 committed:
Add ReaderFrom/WriterTo for Linux

On Linux, we can use `splice` to optimize copies to happen in kernel
when copying to other files.
Since this is already a pipe, this should always work when copying
to/from another file.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

dmcgowan commented on raw_linux.go:
```diff
@@ -0,0 +1,242 @@
+					// splice not supported on kernel
+					atomic.StoreInt32(&spliceSupported, 0)
+					return true
+				case syscall.EINVAL, syscall.EOPNOTSUPP, syscall.EPERM:
```
Would `syscall.EOPNOTSUPP` here indicate the process is unable to perform the syscall regardless of whether the kernel supports it? If this doesn't depend on input would it make sense to treat it like `ENOSYS`?

cpuguy83 replied:
I *think* EOPNOTSUP is more about the transport than the actual system.
For instance, I'd expect `EOPNOTSUP` if the read side is a unix socket which does not support splicing... but I have not tested this... it may actually be worth adding tests for this.

cpuguy83 replied:
Although the specific case may actually be EINVAL...

dmcgowan replied:
That's fair, its not super clear when these standard return values would be returned in this case and better not to set a global value

thaJeztah commented on raw_linux.go:
```diff
@@ -0,0 +1,242 @@
+// +build go1.12
```
These still needed? `go.mod` already has go1.13 as indicator for "minimum supported version"; https://github.com/containerd/fifo/blob/fcd6de7ee90bf3789140ed2f508be259ce8f48ff/go.mod#L3

kzys replied:
+1 for removing the line.

cpuguy83 replied:
Removed.

kzys commented on raw_linux_test.go:
```diff
@@ -0,0 +1,121 @@
+// +build go1.12
```
We can remove this line as well.

cpuguy83 replied:
removed.

cpuguy83 commented:
Updated this, added some new test cases and improved the implementation slightly:

1. EINTR handling was breaking out of the loop instead of just retrying - fixed
2. Before the max amount that could be copied before the new methods would return was `1<<62`... which is a lot... but is not expected since we want to copy until EOF (splice copies 0 bytes), so the internal copy takes a special value, `-1`, to mean copy to EOF. This is important because we have special handling for when the reader passed in to `ReadFrom` is an `*io.LimitedReader` and this was originally mixing "amount to copy" with "amount remaining to copy".

cpuguy83 commented:
I'm going to do some more testing on this as well.

cpuguy83 commented:
FYI, ended up making a new repo to play around with this a bit more, benchmarks, etc: https://github.com/cpuguy83/pipes
Somewhat suprisingly I'm seeing an approx 2x speedup using splice.

I've made some changes to the implementation that make it a bit cleaner which I'll move here as well.

thaJeztah commented:
> Somewhat suprisingly I'm seeing an approx 2x speedup using splice.

Silly question; the benchmark in the readme doesn't show a delta; is that because some option wasn't set?

(Performance improvement sounds great though!)

cpuguy83 commented:
Hmm I'm not sure.
I hadn't used benchstat before...
benchcmp shows the delta.

thaJeztah commented:
@cpuguy83 @kzys @dmcgowan what's the status on this one? Do we want to have this merged and included in a release?

I went looking what changes are in `main` that are not yet released (following https://github.com/containerd/fifo/pull/32#issuecomment-1023065555), so was looking if we wanted to include this PR as well if we would be doing a new release.

kzys commented:
Sorry. I have missed the ping. Let me take a look next week.

samuelkarp commented on raw_linux.go:
```diff
@@ -0,0 +1,251 @@
+		remain = spliceMax
+	}
+
+	// Hear the RawConn Read/Write methods allow us to utilize the go runtime
```
nit
```suggestion
	// Here the RawConn Read/Write methods allow us to utilize the go runtime
```

samuelkarp commented on raw_linux.go:
```diff
@@ -0,0 +1,251 @@
+				case syscall.EINVAL, syscall.EOPNOTSUPP, syscall.EPERM:
+					// In all these cases, there is no data transferred
+					return true
+				case nil:
+					handled = true
+					if n == 0 {
+						// At EOF
+						return true
+					}
+				case unix.EINTR:
+					continue
```
nit: I think it'd be slightly more readable to have an explicit `continue` in here.
```suggestion
				case nil:
					handled = true
					if n == 0 {
						// At EOF
						return true
					}
					continue
				case unix.EINTR:
					continue
```

Also, what do you think about reordering the switch to order successful cases before errors and have all the errors grouped together?  Something like `nil`, `unix.EINTR`, `unix.EAGAIN`, `unix.ENOSYS`, `syscall.EINVAL, syscall.EOPNOTSUPP, syscall.EPERM`, `default`?

samuelkarp commented on raw_linux.go:
```diff
@@ -0,0 +1,251 @@
+	if f.flag&syscall.O_WRONLY > 0 {
+		return 0, ErrRdFrmWRONLY
+	}
+	select {
+	case <-f.opened:
+		return f.writeTo(w)
+	default:
+	}
+
+	select {
+	case <-f.opened:
+		return f.writeTo(w)
```
Same question here about the repeated read from `f.opened`.

samuelkarp commented on raw_linux_test.go:
```diff
@@ -0,0 +1,232 @@
+	data := strings.Repeat("This is a test, this is only a test.", 1000)
+
+	// For these test cases we only call ReadFrom and validate there is no error and the
+	// amouont of data it copied is what we put into it.
```
nit
```suggestion
	// amount of data it copied is what we put into it.
```

samuelkarp reviewed (approved):
A few questions and nits, but otherwise LGTM.

cpuguy83 commented on raw_linux.go:
```diff
@@ -0,0 +1,251 @@
+
+func (f *fifo) readFrom(r io.Reader) (int64, error) {
+	if atomic.LoadInt32(&spliceSupported) != 1 {
+		return copyBuffer(f.file, r)
```
Yeah I think it makes sense to only implement the interfaces where they are supported even though we have a fallback.

dcantah commented:
@cpuguy83 Have any bandwidth to try and push this forward? If not I'm happy to help, this would be sweet to have
